src/hotspot/share/c1/c1_globals.hpp

```diff
@@ -277,7 +277,7 @@
                                                                             \
   develop(intx, NMethodSizeLimit, (64*K)*wordSize,                          \
           "Maximum size of a compiled method.")                             \
-          range(0, max_jint)                                                \
+          range(0, 1*M)                                                     \
```

Comment:
Shouldn't this be defined in platform-specific code, along with a comment explaining why 1MB was chosen?

Reply:
I think a 1MB upper bound is sensible for all platforms (see my other comment below). I'll add a comment explaining the choice, thanks. Please let me know if you think a larger bound is more suitable. In that case, we should also apply @theRealAph's improved aarch64 fix above for `const2reg`.

Reply:
I don't know what a typical nmethod size is for C1, but I can imagine hitting the 1MB limit by doing something like stress testing with increased inlining limits.  Or maybe very large initialization methods in the future thanks to Leyden AOT or computed constants?  So I guess I'm leaning slightly towards fixing the aarch64 issue now.

Reply:
Sorry, that was a bit of a rant. I guess I'm happy that AArch64 will be restricted, if that is the word, to megabyte-long compiled methods, for the aforementioned reasons. Every RISC processor, or at least every processor with fixed-length instructions, will have similar issues.

Reply:
Making the current assumptions explicit by limiting to 1MB on aarch64 seems like a good solution for the moment. I would advise that we also set a default upper limit for all platforms, to avoid other issues in nearly-dead code paths exercised by huge flag values. See, e.g., [JDK-8320302](https://bugs.openjdk.org/browse/JDK-8320302). I've also seen some other potential issues when doing stress testing.